### PR TITLE
Fix "no event handler" in simultaneous blur+removal case

### LINKED ISSUE #26838: "There is no event handler associated with this event" when an element is focused out
### Describe the bug
An error is thrown:
>  Error: System.ArgumentException: There is no event handler associated with this event. EventId: '3'. (Parameter 'eventHandlerId')

when an element with focusout handler is removed from the DOM in a blazor wasm app (both 3.1 and 5.0).

### To Reproduce
1. Clone the https://github.com/Stamo-Gochev/blazor-net5-wasm-focusout repository
2. Set the [TestNet5Wasm.Server](https://github.com/Stamo-Gochev/blazor-net5-wasm-focusout/tree/master/TestNet5Wasm/TestNet5Wasm/Server) as a startup project. 
> **Note:** In order to test the 3.1 version of wasm, use the [TestNet31Wasm.Server](https://github.com/Stamo-Gochev/blazor-net5-wasm-focusout/tree/master/TestNet5Wasm/TestNet31Wasm)
3. Run the app (Ctrl + F5)
4. On the `Index.razor` page, click on `item 2` that has [@onfocusout](https://github.com/Stamo-Gochev/blazor-net5-wasm-focusout/blob/master/TestNet5Wasm/TestNet5Wasm/Client/Pages/Index.razor#L8) handler

> **Note:** The main logic is part of the [Index.razor](https://github.com/Stamo-Gochev/blazor-net5-wasm-focusout/blob/master/TestNet5Wasm/TestNet5Wasm/Client/Pages/Index.razor) page only and the [script.js](https://github.com/Stamo-Gochev/blazor-net5-wasm-focusout/blob/master/TestNet5Wasm/TestNet5Wasm/Client/wwwroot/script.js) file.

### Further technical details
- ASP.NET Core version 3.1, 5.0 RC1
- Include the output of `dotnet --info`
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version

### Additional context
A similar problem is marked to be fixed in multiple cases like:
 https://github.com/dotnet/aspnetcore/issues/17926
https://github.com/dotnet/aspnetcore/issues/21241
https://github.com/dotnet/aspnetcore/issues/21366
https://github.com/dotnet/aspnetcore/issues/21874#issuecomment-632814894

but the problem is still reproducible in both the 3.1 and 5.0 versions.

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Fixes #26838 (a longstanding and tricky bug)

This bug only affects WebAssembly, not Server, due to a difference in how event processing and renderbatch acknowledgement works.

On all hosting platforms, the overall idea for when to dispose event handler ID is this: .NET code should only dispose an event handler after confirmation that the JS-side code has done so. This guarantees that the JS-side code will never raise an event for which .NET has already disposed the handler. .NET knows it's safe to do this once it receives an acknowledgement for the renderbatch that contains the event disposal.

* On Server, all messages from JS to .NET are async and arrive in a single queue. Whenever we process a renderbatch acknowledgement, we know that any preceding event notifications have already started being dispatched (and hence have already captured a reference to the handler they will need) because of that message queue.
* On WebAssembly, there isn't (or wasn't) really a true notion of renderbatch acknowledgements, because we know that batches are always applied synchronously, hence it's valid to just treat all batches as immediately and synchronously acknowledged as soon as we've called the "update DOM" JS method.
  * However, there's an edge case where this is false. In order for WebAssembly to guarantee that events aren't nested inside each other on the stack (which is not possible on Server because of how they arrive as separate async events), it has a system of "deferring" an incoming event if one is already on the call stack, until we pop back off to the top of the call stack. Imagine the deferred event has ID=X. In the unlikely event that the event already on the call stack causes a render that removes X, the fact that we treat the batch as synchronously acknowledged means that we then proceed to dispose X before processing the deferred event. Crash.

To fix this, my overall goal is to bring WebAssembly and Server call ordering more in sync with each other, both to fix this one issue and to reduce the chances of other similar issues in the future. So what I've done is to make WebAssembly no longer treat renderbatch acknowledgements as always completed synchronously, and instead:

 * If there are no deferred events waiting to be processed, behave like before and just return `Task.CompletedTask` (so the ack is indeed still sync - no perf lost)
 * But in the edge case where there is a deferred event, chain the acknowledgement onto a `Task` that fires once the currently-last-in-queue deferred event starts processing. This is as if there was a single queue of incoming messages, and we inserted the renderbatch acknowledgement at the tail of it. Which is exactly like Blazor Server.

So in summary, the WebAssembly/Server ordering inconsistency edge case is gone, and hence so is the bug with `blur`. But it won't have any impact on any event processing other than these deferred ones.

Thanks for reading my long message. If it made no sense, just click *Approve* and move on with your life.

### CODE REVIEW

File: src/Components/test/testassets/BasicTestApp/FocusEventComponent.razor [hunk 1]
```diff
@@ -12,40 +12,59 @@
...
 @code {
-
+    bool showButtonThatDisappearsWhenClicked = true;
+    bool buttonReceivedFocusOut;
     string message;
 
     void OnFocus(FocusEventArgs e)
     {
         message += "onfocus,";
-        StateHasChanged();
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/FocusEventComponent.razor:26]: Removing these `StateHasChanged` calls is a bit of unrelated cleanup. They're just not needed.

**Review by javiercn — APPROVED** [review]
Looks great!

**Comment by Stamo-Gochev** [comment]
@SteveSandersonMS I see that the milestone for this fix is set to [6.0-preview4](https://github.com/dotnet/aspnetcore/milestone/145), but the original bug was reported for versions 3.1 and 5.0. Will it be ported to 5.0 (as wasm 3.2 is now out of support)? 

According to the [support policy](https://dotnet.microsoft.com/platform/support/policy/dotnet-core) the end support date of 5.0 is "3 months after .NET 6 release (around February 2022)", so it will be almost a year since the bug fix can be used for projects targeting 5.0. Is there any chance to add it to 5.0 as well?
